modules/presence/presence.c

```diff
@@ -897,8 +903,10 @@ static struct mi_root* mi_list_shtable(struct mi_root* cmd, void* param)
 		}
 		lock_release(&subs_htable[i].lock);
 
-		if ((j % 50) == 0)
+		if (j >= 50) {
 			flush_mi_tree(rpl_tree);
+			j = 0;
+		}
```

Comment:
I think this change does not bring any modification in behavior - correct me if I'm wrong.

Reply:
Well. This was the direct cause of our problems:

- mi_list_shtable was called
- j = 0
- lock is attained on subs_htable[0]
- no matches are found, j is still 0
- unlock op subs_htable[0]
- j%50 == 0, thus we flush even though mi_print_shtable_record was not added

Alternately, we could go from j=49 to j=52 inside the for, and then we wouldn't flush, even though it is time.

Because the increment of j is done inside the for(), we're not guaranteed to hit this flush-condition exactly at j%50==0. Therefore the change so we flush at around 50 or more.

Reply:
Hi @wdoekes , if this is the issue, maybe the easiest fix will be to start with `j=1;`

Reply:
That would fix problem 1:
> j%50 == 0, thus we flush even though mi_print_shtable_record was not added

But it would not fix problem 2:
> Alternately, we could go from j=49 to j=52 inside the for, and then we wouldn't flush, even though it is time.

Because there may be several matches inside the for(), causing the `if((j%50)==0)` not to be reached at every increment of j.

Reply:
The alternate solution was to move the `if()` to inside the `for()`, but that would also put it inside the locked section, which I wanted to avoid.

Reply:
oh, thanks for your patience here, I see your point. I was a bit confused as I wasn't expecting this. In other modules, like [usrloc](https://github.com/OpenSIPS/opensips/blob/2.4/modules/usrloc/ul_mi.c#L441)  or [dialog](https://github.com/OpenSIPS/opensips/blob/2.4/modules/dialog/dlg_hash.c#L1426), the flush is done under the lock. If acceptable for you, I would prefer to keep the same approach here too (for consistency purposes). Of course, starting with j = 1 :)

Reply:
No problem. Force pushed. It now looks like this:
https://github.com/OpenSIPS/opensips/pull/1614/commits/fd3a6e4a71674c626c8b7f97f9892d08e48d0f7f#diff-a09dfe59b017a1cfe2ef7b05b4f97203R900